frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/backgroundables/RenameBackgroundable.java

```diff
@@ -0,0 +1,91 @@
+import com.intellij.openapi.vcs.VcsNotifier;
+import git4idea.repo.GitRepository;
+import lombok.experimental.ExtensionMethod;
+import lombok.val;
+import org.checkerframework.checker.guieffect.qual.UIEffect;
+
+import com.virtuslab.binding.RuntimeBinding;
+import com.virtuslab.branchlayout.api.BranchLayout;
+import com.virtuslab.branchlayout.api.readwrite.IBranchLayoutWriter;
+import com.virtuslab.gitmachete.frontend.file.MacheteFileWriter;
+import com.virtuslab.gitmachete.frontend.vfsutils.GitVfsUtils;
+import com.virtuslab.qual.guieffect.UIThreadUnsafe;
+
+@ExtensionMethod(GitVfsUtils.class)
+public class RenameBackgroundable extends Task.Backgroundable {
+
+  protected final Project project;
+  protected final GitRepository gitRepository;
+  protected final BranchLayout branchLayout;
+  protected final Runnable renameRunnable;
+  protected final String currentBranchName;
+  protected final String newBranchName;
+
+  public RenameBackgroundable(
+      GitRepository gitRepository,
+      BranchLayout branchLayout,
+      Runnable renameRunnable,
+      String currentBranchName,
+      String newBranchName) {
+    super(gitRepository.getProject(), getString("action.GitMachete.RenameBackgroundable.task-title"));
+    this.project = gitRepository.getProject();
+    this.gitRepository = gitRepository;
+    this.branchLayout = branchLayout;
+    this.renameRunnable = renameRunnable;
+    this.currentBranchName = currentBranchName;
+    this.newBranchName = newBranchName;
+  }
+
+  @Override
+  @UIThreadUnsafe
+  public void run(ProgressIndicator indicator) {
+    Path macheteFilePath = gitRepository.getMacheteFilePath();
+
+    val newBranchLayout = branchLayout.rename(currentBranchName, newBranchName);
+    val branchLayoutWriter = RuntimeBinding.instantiateSoleImplementingClass(IBranchLayoutWriter.class);
+
+    runWriteActionOnUIThread(() -> {
+      MacheteFileWriter.writeBranchLayout(
+          macheteFilePath,
+          branchLayoutWriter,
+          newBranchLayout,
+          /* backupOldLayout */ true,
+          /* requestor */ this);
+
+      renameRunnable.run();
+
+      // `renameRunnable` may perform some sneakily-asynchronous operations (e.g. renameBranch).
+      // The high-level method used within the runnable does not allow us to schedule the tasks after them.
+      // (Stepping deeper is not an option since we would lose some important logic or become very dependent on the internals of git4idea).
+      // Hence, we wait for the creation of the branch (with exponential backoff).
+      //
+      // Theoretically we could not wait for the branch creation.
+      // However, in would result in a premature execution of GitMacheteRepositoryUpdateBackgroundable,
+      // and eventually with a waring about a branch listed in the machete file that does not exist.
+      // GitMacheteRepositoryUpdateBackgroundable will wait for this backgroundable to finish.
+      waitForCreationOfLocalBranch(gitRepository, newBranchName);
```

Comment:
Whooops wait

Reply:
This is executed on UI thread? 😱

Reply:
Nooooooooooo

Reply:
It wasn't detected by the ArchUnit test coz it can't deal with `@UI Runnable`s ☹️

Reply:
I think so. it's in `runWriteActionOnUIThread`

Reply:
I do believe it might prevent race conditions with the warning dialog indeed 😅 but at the expense of blocking UI thread... folks might soon report freezes like #1457 🥶

Reply:
oh, that's why it's working and that's bad 😢

Reply:
I'd reconsider the solution with:
1. slide out from layout
2. rename in git
3. overwrite layout with the original but with renamed branch

Reply:
and I see an issue already! our slide in suggestion my appear while renaming the current branch xd

Reply:
> I'd reconsider the solution with:

Or maybe even better/simpler! How about just putting some lock (a simple boolean/AtomicBoolean?) on `com.virtuslab.gitmachete.frontend.ui.api.table.BaseEnhancedGraphTable#queueRepositoryUpdateAndModelRefresh`? To be acquired when rename action starts, and released once completed (in `finally`, so that it's always released, even if rename fails)